internal/driver/flamegraph.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package driver
+
+import (
+	"encoding/json"
+	"html/template"
+	"net/http"
+
+	"github.com/google/pprof/internal/report"
+)
+
+type flameGraphNode struct {
+	Name     string
+	Value    int64
+	Children map[string]*flameGraphNode
+}
+
+func (n *flameGraphNode) add(stackPtr *[]string, index int, value int64) {
+	n.Value += value
+	if index >= 0 {
+		head := (*stackPtr)[index]
+		childPtr, ok := n.Children[head]
+		if !ok {
+			childPtr = &(flameGraphNode{head, 0, make(map[string]*flameGraphNode)})
+			n.Children[head] = childPtr
+		}
+		childPtr.add(stackPtr, index-1, value)
+	}
+}
+
+func (n *flameGraphNode) MarshalJSON() ([]byte, error) {
+	v := make([]flameGraphNode, 0, len(n.Children))
+	for _, value := range n.Children {
+		v = append(v, *value)
+	}
+
+	return json.Marshal(&struct {
```

Comment:
It would be better to format the values on the server side and have the client use that formatted value for the presentation. There are formatting helpers in the measurement package, using this would make the view more unified with the other views and it handles things like memory units.

TextItem might be a good structure to reuse - https://github.com/google/pprof/blob/f0300aaaf05b5fe9bec4a79ff111f9578c03e533/internal/report/report.go#L710, the top view in the webui already uses that.

Reply:
The d3 plugin (and all other d3-hierarchy based plugins) only take a specific format. Sending a different format to the browser means more JavaScript code to convert that format to JSON, unless there's a formatting helper that matches the d3-hierarchy format.